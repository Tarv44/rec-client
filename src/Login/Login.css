.login-form {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 0 12%;
    margin-top: 50%;
}

.login-form h2 {
    align-self: center;
}

.login-form > * {
    margin-bottom: 10px;
}

.login-form > *:last-child {
    margin-bottom: 0;
}

.login-form button {
    width: 60%;
    margin-top: 10px;
}

.form-group {
    width: 100%;
    text-align: right;
    
}

.form-group input {
    width: 50%;
    margin-left: 5px;
}

.error {
    text-align: center;
    color: red;
}