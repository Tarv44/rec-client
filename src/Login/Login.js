import { Component } from 'react'
import ValidationError from '../ValidationError'
import store from '../dummy-store'
import RotationContext from '../RotationContext'
import './Login.css'

export default class Login extends Component {
    static contextType = RotationContext

    constructor(props) {
        super(props)
        this.state = {
            loginFail: {
                message: '',
                failed: false
            },
            username: {
                value: '',
                touched: false
            },
            password: {
                value: '',
                touched: false
            }
        }
    }

    updateUsername(username) {
        this.setState({ username: { value: username, touched: true } })
    }

    updatePassword(password) {
        this.setState({ password: { value: password, touched: true } })
    }

    validateUsername() {
        const username = this.state.username.value.trim()
        if (username.length === 0) {
            return 'Username is required'
        } 
    }

    validatePassword() {
        const password = this.state.password.value.trim()
        if (password.length === 0) {
            return 'Username is required'
        } 
    }

    handleLoginSubmit(e) {
        e.preventDefault()
        const username = this.state.username.value.trim().toLowerCase()
        const password = this.state.password.value.trim()
        //This function will eventually make a GET request to database to verify account.
        //For now, it verifies account in dummy store and responds accordingly
        const userIdx = store.users.findIndex(user => user.username.toLowerCase() === username)
        if (userIdx === -1) {
            this.setState({
                loginFail: {
                    failed: true,
                    message: 'Username not found.'
                }
            })
        } else {
            if (password !== store.users[userIdx].password) {
                this.setState({
                    loginFail: {
                        message: 'Incorrect password.',
                        failed: true
                    }
                })
            } else {
                const userId = store.users[userIdx].id
                this.context.updateUser(this.state.username.value, userId)
                this.setState({
                    loginFail: {
                        message: '',
                        failed: false
                    },
                    username: {
                        value: '',
                        touched: false
                    },
                    password: {
                        value: '',
                        touched: false
                    }
                })
                this.props.history.push('/dashboard')
            }
        }
    }

    render() {
        return (
            <form className='login-form' onSubmit={e => this.handleLoginSubmit(e)}>
                <h2>Login</h2>
                {this.state.loginFail && <p className='error'>{this.state.loginFail.message}</p>}
                <div className='form-group'>
                    <label htmlFor='username'>Username:</label>
                    <input type='text' name='username' id='username' onChange={e => this.updateUsername(e.target.value)}/>
                    {this.state.username.touched && <ValidationError message={this.validateUsername()}/>}
                </div>
                <div className='form-group'>
                    <label htmlFor='password'>Password:</label>
                    <input type='password' name='password' id='password' onChange={e => this.updatePassword(e.target.value)}/>
                    {this.state.password.touched && <ValidationError message={this.validatePassword()}/>}
                </div>
                <button type='submit'>Login</button>
            </form>
        )
    }
}