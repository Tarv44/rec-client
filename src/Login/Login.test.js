import React from 'react'
import ReactDom from 'react-dom'
import renderer from 'react-test-renderer'
import Login from './Login'
import { BrowserRouter } from 'react-router-dom'


describe('----- Login -----', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div')
        ReactDom.render((
            <BrowserRouter>
                <Login />
            </BrowserRouter> 
        ), div)
        ReactDom.unmountComponentAtNode(div);
    })
    it('renders the UI as expected', () => {
        const comp = renderer
            .create(
                <BrowserRouter>
                    <Login />
                </BrowserRouter> 
            )
            .toJSON()
        expect(comp).toMatchSnapshot()
    })
})