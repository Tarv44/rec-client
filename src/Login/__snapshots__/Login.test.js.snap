// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`----- Login ----- renders the UI as expected 1`] = `
<main
  className="signup-login"
>
  <form
    autoComplete="off"
    className="login-form"
    onSubmit={[Function]}
  >
    <h2>
      Login
    </h2>
    <div
      className="form-group"
    >
      <label
        htmlFor="email"
      >
        Email
      </label>
      <input
        id="email"
        name="email"
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <input
        id="password"
        name="password"
        onChange={[Function]}
        type="password"
      />
    </div>
    <button
      type="submit"
    >
      Login
    </button>
    <p>
      Don't have an account?
    </p>
    <a
      aria-current={null}
      className="login-here"
      href="/signup"
      onClick={[Function]}
    >
      Signup here.
    </a>
  </form>
</main>
`;
